Library/Homebrew/command.rb

```diff
@@ -0,0 +1,133 @@
+      @parent_option_name = parent_option_name
+    end
+
+    def error_message(argv_tokens = @argv_tokens)
+      invalid_options =
+        argv_tokens.select { |arg| /^--/ =~ arg }
+                   .reject { |arg| @valid_options.map { |opt| opt[:option_name] }.include?(arg) }
+      return if invalid_options.empty?
+      "Invalid option(s) provided: #{invalid_options.join " "}"
+    end
+
+    def check_for_errors
+      return if error_message.nil?
+      odie <<-EOS.undent
+        #{error_message}
+        Correct usage:
+        #{@help_output}
+      EOS
+    end
+
+    def generate_documentation
+      docs = Documentation.new(@command_name, @description, @valid_options)
+      @help_output = docs.help_output
+      @man_output = docs.man_output
+    end
+  end
+
+  class Documentation
+    attr_reader :help_output, :man_output
+
+    def initialize(command_name, description, valid_options)
+      root_option_names = root_option_names(valid_options)
+      option_str = root_option_names.map do |ro|
+        option_string(ro, valid_options)
+      end.join(" ").gsub(/\s+/, " ")
+      desc_str = root_option_names.map do |ro|
+        desc_string(ro, valid_options)
+      end.join("\n\s\s\s\s")
+
+      help_lines = "\s\s" + <<-EOS.undent
+        * `#{command_name}` #{option_str}:
+            #{description}
+
+            #{desc_str}
+      EOS
+                                  .chop
+      @man_output = help_lines
+      help_lines = help_lines.split("\n")
+      help_lines.map! do |line|
+        line
+          .sub(/^  \* /, "#{Tty.bold}brew#{Tty.reset} ")
+          .gsub(/`(.*?)`/, "#{Tty.bold}\\1#{Tty.reset}")
+          .gsub(/<(.*?)>/, "#{Tty.underline}\\1#{Tty.reset}")
+      end
+      @help_output = help_lines.join("\n")
+    end
+
+    def option_string(option, valid_options)
+      hash = valid_options.find { |opt| opt[:option_name] == option }
+      child_options = hash[:child_option_names]
+      output = "[`#{option}`]"
+      unless child_options.empty?
+        childs_str = child_options.map do |co|
+          option_string(co, valid_options)
+        end.join(" ")
+        output = output.gsub(/`#{option}`/, "\\0 #{childs_str}")
+      end
+      output
+    end
+
+    def desc_string(option, valid_options, parent_option_present = false)
+      hash = valid_options.find { |opt| opt[:option_name] == option }
+      desc = hash[:desc]
+      child_options = hash[:child_option_names]
+
+      output = <<-EOS.undent
+        `#{option}`, #{desc}
+      EOS
+      if parent_option_present
+        output = output.gsub(/`#{option}`/, 'With \\0')
+      else
+        output = output.gsub(/`#{option}`/, 'If \\0 is passed')
+      end
+      unless child_options.empty?
+        childs_str = child_options.map do |co|
+          desc_string(co, valid_options, true)
+        end.join("\s\s\s\s")
+        output = <<-EOS.undent
+          #{output}\s\s\s\s#{childs_str}
+        EOS
+                       .chop
```

Comment:
No longer needed. Defined this as a single one line string rather than a `EOS` block